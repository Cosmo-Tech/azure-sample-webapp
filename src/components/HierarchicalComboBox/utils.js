// Copyright (c) Cosmo Tech.
// Licensed under the MIT license.

function addFormattedOption (dataObject, maxCharLength, separator, depth, formattedList) {
  const option = { ...dataObject };
  option.fullName = dataObject.name;
  const subStringLength = maxCharLength / 2;
  option.id = dataObject.id;
  (maxCharLength === -1 || dataObject.name.length <= maxCharLength)
    ? option.name = dataObject.name
    : option.name = dataObject.name.substring(0, subStringLength) + separator + dataObject.name.substring(dataObject.name.length - subStringLength);
  option.depth = depth;
  formattedList.push(option);
}

function findDepthValue (dataList, currentObject, depth) {
  if (currentObject !== undefined) {
    if (currentObject.parentId !== null) {
      const parent = dataList.find(element => element.id === currentObject.parentId);
      return findDepthValue(dataList, parent, depth + 1);
    }
  }
  return depth;
}

export const getFormattedOptionsList = (formattedList, dataList, depth, separator, maxCharLength) => {
  if (dataList !== undefined && dataList.length > 0) {
    for (const dataObject of dataList) {
      const depth = findDepthValue(dataList, dataObject, 0);
      addFormattedOption(dataObject, maxCharLength, separator, depth, formattedList);
    }
  }
};
