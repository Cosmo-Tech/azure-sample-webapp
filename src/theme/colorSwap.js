// Copyright (c) Cosmo Tech.
// Licensed under the MIT license.

export const colorSwap = {
  primary: {
    0: '#000000',
    10: '#001d36',
    20: '#003258',
    25: '#003d69',
    30: '#00497c',
    35: '#00558f',
    40: '#0061a3',
    50: '#317bbe',
    60: '#5095d9',
    70: '#6daff6',
    80: '#9dcaff',
    90: '#d1e4ff',
    95: '#e9f1ff',
    98: '#f8f9ff',
    99: '#fdfcff',
    100: '#ffffff',
  },
  secondary: {
    0: '#000000',
    10: '#2a1800',
    20: '#462b00',
    25: '#553500',
    30: '#643f00',
    35: '#734900',
    40: '#845400',
    50: '#a56b00',
    60: '#c78200',
    70: '#e79c24',
    80: '#ffb958',
    90: '#ffddb5',
    95: '#ffeedd',
    98: '#fff8f4',
    99: '#fffbff',
    100: '#ffffff',
  },
  tertiary: {
    0: '#000000',
    10: '#251432',
    20: '#3b2948',
    25: '#463453',
    30: '#52405f',
    35: '#5e4b6c',
    40: '#6a5778',
    50: '#846f92',
    60: '#9f89ad',
    70: '#baa3c8',
    80: '#d6bee5',
    90: '#f2daff',
    95: '#fbecff',
    98: '#fff7fd',
    99: '#fffbff',
    100: '#ffffff',
  },
  error: {
    0: '#000000',
    10: '#410002',
    20: '#690005',
    25: '#7e0007',
    30: '#93000a',
    35: '#a80710',
    40: '#ba1a1a',
    50: '#de3730',
    60: '#ff5449',
    70: '#ff897d',
    80: '#ffb4ab',
    90: '#ffdad6',
    95: '#ffedea',
    98: '#fff8f7',
    99: '#fffbff',
    100: '#ffffff',
  },
  neutral: {
    0: '#000000',
    10: '#1a1c1e',
    20: '#2f3033',
    25: '#3a3b3e',
    30: '#45474a',
    35: '#515255',
    40: '#5d5e61',
    50: '#76777a',
    60: '#909094',
    70: '#ababae',
    80: '#c6c6ca',
    90: '#e2e2e6',
    95: '#f1f0f4',
    98: '#faf9fc',
    99: '#fdfcff',
    100: '#ffffff',
  },
  neutralVariant: {
    0: '#000000',
    10: '#171c22',
    20: '#2c3137',
    25: '#373c42',
    30: '#42474e',
    35: '#4e535a',
    40: '#5a5f66',
    50: '#73777f',
    60: '#8d9199',
    70: '#a7abb4',
    80: '#c3c7cf',
    90: '#dfe2eb',
    95: '#edf1f9',
    98: '#f8f9ff',
    99: '#fdfcff',
    100: '#ffffff',
  },
};
