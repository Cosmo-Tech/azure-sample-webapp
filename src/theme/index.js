// Copyright (c) Cosmo Tech.
// Licensed under the MIT license.

import { paletteLight, pictureLight, gridLight, paletteDark, pictureDark, gridDark } from './custom';
import { createTheme, adaptV4Theme } from '@mui/material/styles';
import overrides from './overrides';

export { paletteLight, pictureLight, gridLight, paletteDark, pictureDark, gridDark, overrides };
export const getTheme = (isDarkTheme) =>
  createTheme(
    adaptV4Theme({
      palette: isDarkTheme ? paletteDark : paletteLight,
      picture: isDarkTheme ? pictureDark : pictureLight,
      grid: isDarkTheme ? gridDark : gridLight,
      overrides,
      components: {
        MuiPaper: {
          styleOverrides: { root: { backgroundImage: 'unset' } },
        },
      },
    })
  );
