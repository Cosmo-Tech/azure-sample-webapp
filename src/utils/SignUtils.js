// Copyright (c) Cosmo Tech.
// Licensed under the MIT license.

import { Auth } from '@cosmotech/core';

export const signIn = async (callback) => {
  if (Auth.isAsync()) {
    return Auth.isUserSignedIn(callback);
  } else {
    return Auth.isUserSignedIn();
  }
};
