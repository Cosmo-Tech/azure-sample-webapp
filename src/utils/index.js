// Copyright (c) Cosmo Tech.
// Licensed under the MIT license.

export { ConfigUtils } from './ConfigUtils';
export { DatasetsUtils } from './DatasetsUtils';
export { TranslationUtils } from './TranslationUtils';
export { SolutionsUtils } from './SolutionsUtils';
export { ScenarioParametersUtils } from './ScenarioParametersUtils';
