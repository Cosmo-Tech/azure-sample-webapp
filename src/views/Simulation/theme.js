// Copyright (c) Cosmo Tech.
// Licensed under the MIT license.
export const simulationTheme = {
  dark: {
    mode: 'dark',
    graph: {
      link: '#f5f3f3',
      hiddenLink: '#636363',
      redLine: '#df3537',
      grayLine: '#b9bac0',
      stockLevel0Fill: '#20363D',
      stockLevel1Fill: '#DF3537',
      stockLevel0Border: '#48C0DB52',
      stockLevel1Border: '#DF3537',
      productionResourceFill: '#20363D',
      productionResourceBorder: '#48C0DB52',
      productionResourceIcon: '#40B8D4',
      factoryIcon: '#b9bac0',
      operationsCountBadge: '#2F6A79',
    },
    minimap: {
      background: '#121212',
      stock: '#48c0db',
      stockShortage: '#df3537',
      productionResource: '#48c0db',
      productionResourceBottleneck: '#df3537',
    },
    button: {
      color: '#F7F7F8',
      background: '#232323',
      icon: {
        currentMode: '#FFC669',
        buttonMode: '#F7F7F8',
      },
      hoverBackground: '#1C1C20',
    },
    TimelineControls: {
      background: '#1C1C20',
      progressFill: '#252529',
    },
    map: {
      country: '#333333',
      border: '#757272',
    },
    text: '#ffffff',
    background: '#121212',
  },
  light: {
    mode: 'light',
    graph: {
      link: '#3322f2',
      hiddenLink: '#160e73',
      redLine: '#df3537',
      grayLine: '#b9bac0',
      stockLevel0Fill: '#20363D',
      stockLevel1Fill: '#DF3537',
      stockLevel0Border: '#48C0DB52',
      stockLevel1Border: '#DF3537',
      productionResourceFill: '#20363D',
      productionResourceBorder: '#48C0DB52',
      productionResourceIcon: '#40B8D4',
      factoryIcon: '#606061ff',
      operationsCountBadge: '#2F6A79',
    },
    minimap: {
      background: '#ebebebff',
      stock: '#034757ff',
      stockShortage: '#e00004ff',
      productionResource: '#034757ff',
      productionResourceBottleneck: '#e00004ff',
    },
    button: {
      background: '#ebebebff',
      icon: {
        currentMode: '#ffd38bff',
        buttonMode: '#1C1C20',
      },
      color: '#1C1C20',
      hoverBackground: '#979797ff',
    },
    TimelineControls: {
      background: '#ebebebff',
      progressFill: '#701c1cff',
    },
    text: '#000000',
    background: '#FAFAFA',
  },
};
