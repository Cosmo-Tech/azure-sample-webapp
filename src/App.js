// Copyright (c) Cosmo Tech.
// Licensed under the MIT license.

import React, { useCallback, useEffect, useMemo, useState } from 'react';
import { useTranslation } from 'react-i18next';
import { useIdleTimer } from 'react-idle-timer';
import { Auth } from '@cosmotech/core';
import { ThemeProvider, CssBaseline } from '@material-ui/core';
import './assets/scss/index.scss';
import './services/config/Auth';
import { TABS } from './AppLayout';
import Loading from './views/Loading';
import './services/AppInsights';
import { AUTH_STATUS } from './state/commons/AuthConstants';
import { SESSION_INACTIVITY_TIMEOUT } from './services/config/FunctionalConstants';
import { SessionTimeoutDialog } from './components/SessionTimeoutDialog/SessionTimeoutDialog';
import { getTheme } from './theme';
import AppRoutes from './AppRoutes';
import { STATUSES } from './state/commons/Constants';
import { useApp } from './AppHook';

const SESSION_TIMEOUT_PROMPT_DELAY_IN_SECONDS = 30;

const App = () => {
  const { t } = useTranslation();
  document.title = t('commoncomponents.text.application.title', 'Cosmo Tech Web Application Sample');

  const [applicationStatus, authStatus, isDarkTheme, getAllInitialData, setApplicationStatus, logIn, logOut] = useApp();

  const theme = React.useMemo(() => getTheme(isDarkTheme), [isDarkTheme]);

  const isAuthenticated = authStatus === AUTH_STATUS.AUTHENTICATED || authStatus === AUTH_STATUS.DISCONNECTING;
  const isConnecting = authStatus === AUTH_STATUS.CONNECTING;

  useEffect(() => {
    if (isAuthenticated) {
      getAllInitialData();
    } else {
      setApplicationStatus(STATUSES.IDLE);
    }
  }, [getAllInitialData, isAuthenticated, setApplicationStatus]);

  useEffect(() => {
    // Check if the user is already signed-in
    if (authStatus === AUTH_STATUS.UNKNOWN) {
      logIn();
    }
  }, [authStatus, logIn]);

  const [isSessionTimeoutDialogOpen, setIsSessionTimeoutDialogOpen] = useState(false);
  const closeSessionTimeoutDialog = () => {
    idleTimer?.reset();
    setIsSessionTimeoutDialogOpen(false);
  };
  const sessionTimeoutLabels = {
    body: t(
      'commoncomponents.dialog.sessionTimeout.body',
      'You have been idle for a long time. You will be automatically disconnected in a few seconds...'
    ),
    cancel: t('commoncomponents.dialog.sessionTimeout.cancel', 'Cancel'),
    logOut: t('commoncomponents.dialog.sessionTimeout.logOut', 'Sign out'),
    loggingOut: t('commoncomponents.dialog.sessionTimeout.loggingOut', 'Signing out...'),
    title: t('commoncomponents.dialog.sessionTimeout.title', 'Session timeout'),
  };
  const getRemainingTimeLabel = (seconds) =>
    t('views.accessdenied.signouttimeout', 'You will be automatically signed out in {{seconds}} seconds...', {
      seconds: seconds,
    });

  let idleTimer = {};

  const onIdle = async () => {
    const authenticated = await Auth.isUserSignedIn();
    if (authenticated) {
      setIsSessionTimeoutDialogOpen(true);
    }
  };

  const timeout = 1000 * 60 * SESSION_INACTIVITY_TIMEOUT;
  idleTimer = useIdleTimer({ onIdle, timeout });

  const isLoading = useMemo(
    () =>
      applicationStatus !== STATUSES.ERROR &&
      (applicationStatus === STATUSES.LOADING || applicationStatus === STATUSES.IDLE),
    [applicationStatus]
  );

  const getAppContent = useCallback(() => {
    if (isConnecting) {
      return <div className="spinner-border text-success" role="status" />;
    }

    if (isAuthenticated && isLoading) {
      return <Loading />;
    }

    return (
      <AppRoutes authenticated={isAuthenticated} authorized={applicationStatus === STATUSES.SUCCESS} tabs={TABS} />
    );
  }, [isConnecting, isAuthenticated, isLoading, applicationStatus]);

  return (
    <>
      <ThemeProvider theme={theme}>
        <CssBaseline />
        <SessionTimeoutDialog
          getRemainingTimeLabel={getRemainingTimeLabel}
          labels={sessionTimeoutLabels}
          open={isSessionTimeoutDialogOpen}
          onClose={closeSessionTimeoutDialog}
          onLogOut={logOut}
          timeout={SESSION_TIMEOUT_PROMPT_DELAY_IN_SECONDS}
        />
        {getAppContent()}
      </ThemeProvider>
    </>
  );
};

export default App;
