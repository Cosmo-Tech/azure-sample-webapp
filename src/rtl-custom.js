// Copyright (c) Cosmo Tech.
// Licensed under the MIT license.

import React, { Suspense } from 'react';
import { render as rtlRender } from '@testing-library/react';
import { ThemeProvider } from '@material-ui/styles';
import PropTypes from 'prop-types';
import { BrowserRouter } from 'react-router-dom';
import { Provider } from 'react-redux';
import { createStore } from 'redux';
import reducer from './state/reducers/rootReducer';
import theme from './theme';

const customRender = (ui, {
  initialStateForStore,
  store = createStore(reducer, initialStateForStore),
  ...renderOptions
} = {}) => {
  const AllTheProviders = ({ children }) => {
    return (
        <Suspense fallback="loading">
          <BrowserRouter>
            <Provider store={store}>
              <ThemeProvider theme={theme}>
                  {children}
              </ThemeProvider>
            </Provider>
          </BrowserRouter>
        </Suspense>
    );
  };
  AllTheProviders.propTypes = {
    children: PropTypes.element
  };
  return rtlRender(ui, { wrapper: AllTheProviders, ...renderOptions });
};

// re-export everything
export * from '@testing-library/react';

// override render method
export { customRender as render };
