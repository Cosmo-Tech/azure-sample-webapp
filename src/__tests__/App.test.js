// Copyright (c) Cosmo Tech.
// Licensed under the MIT license.

import React, { useState as useStateMock } from 'react';
import App from '../App';
import { render } from '../rtl-custom';
import configureStore from 'redux-mock-store';
import renderer from 'react-test-renderer';
import { signIn } from '../utils/SignUtils';
import { Provider } from 'react-redux';
import theme from '../theme';
import { BrowserRouter } from 'react-router-dom';
import { ThemeProvider } from '@material-ui/styles';

const initialStore = {
  scenario: {
    list: {
      data: [],
      status: 'IDLE'
    },
    current: {
      data: null,
      status: 'IDLE'
    }
  },
  application: {
    status: 'IDLE'
  }
};
const mockStore = configureStore([]);

jest.mock('react', () => ({
  ...jest.requireActual('react'),
  useState: jest.fn()
}));

jest.mock('../utils/SignUtils');

describe('App test suite with state default value', () => {
  const setState = jest.fn();

  beforeEach(() => {
    useStateMock
      .mockImplementation(init => [init, setState]);

    signIn.mockImplementation(() => Promise.resolve(true));
  });

  // TODO Fix following tests
  // eslint-disable-next-line jest/no-disabled-tests
  it.skip('should render correclty the loading auth', () => {
    const elem = renderer.create(<App />).toJSON();
    expect(elem).toMatchSnapshot();
  });

  // eslint-disable-next-line jest/expect-expect
  // eslint-disable-next-line jest/no-disabled-tests
  it.skip('renders App loading', () => {
    const { getByText } = render(<App />, { initialStateForStore: initialStore, store: mockStore(initialStore) });
    getByText('views.common.text.loading');
  });
});

// TODO Fix following tests
// Errors thrown:
// - The above error occurred in the <ForwardRef(Modal)> component
// - The above error occurred in the <ForwardRef(Tabs)> component
// See thread : https://stackoverflow.com/questions/67227163/material-ui-react-test-renderer-react-router-tabs-tab-component-and-modal
describe('App test suite with loading value to false', () => {
  const setState = jest.fn();

  beforeEach(() => {
    useStateMock
      .mockImplementationOnce(init => [init, setState])
      .mockImplementationOnce(init => [init, setState])
      .mockImplementationOnce(() => [false, setState]);

    signIn.mockImplementation(() => Promise.resolve(true));
  });

  // eslint-disable-next-line jest/no-disabled-tests
  it.skip('should render correclty the loading auth', () => {
    const elem = renderer.create(
        <BrowserRouter>
          <Provider store={mockStore(initialStore)}>
            <ThemeProvider theme={theme}>
              <App/>
            </ThemeProvider>
          </Provider>
        </BrowserRouter>).toJSON();
    expect(elem).toMatchSnapshot();
  });

  // eslint-disable-next-line jest/expect-expect
  // eslint-disable-next-line jest/no-disabled-tests
  it.skip('renders App loading', () => {
    const { getByText } = render(<App />, { initialStateForStore: initialStore, store: mockStore(initialStore) });
    getByText('views.common.text.loading');
  });
});
