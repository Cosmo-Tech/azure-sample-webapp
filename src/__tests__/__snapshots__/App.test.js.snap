// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App test suite with state default value should render correclty the loading auth 1`] = `
<div
  className="spinner-border text-success"
  role="status"
>
  <span
    className="sr-only"
  >
    views.common.text.loading
  </span>
</div>
`;
